controllers/controllers/clusters/vsphere.go

```diff
@@ -2,19 +2,153 @@ package clusters
+func NewVSphereReconciler(client client.Client, log logr.Logger, validator *vsphere.Validator, defaulter *vsphere.Defaulter) *VSphereClusterReconciler {
+	capiResourceFetcher := resource.NewCAPIResourceFetcher(client, log)
+	return &VSphereClusterReconciler{
+		VSphereReconciler: VSphereReconciler{
+			Client:    client,
+			Log:       log,
+			Validator: validator,
+			Defaulter: defaulter,
+		},
+		providerClusterReconciler: &providerClusterReconciler{},
+		capiResourceFetcher:       capiResourceFetcher,
+	}
 }
 
-func (v *VSphereReconciler) Reconcile(ctx context.Context, cluster *anywherev1.Cluster) (reconciler.Result, error) {
+func (v *VSphereReconciler) VsphereCredentials(ctx context.Context) (*apiv1.Secret, error) {
+	secret := &apiv1.Secret{}
+	secretKey := client.ObjectKey{
+		Namespace: "eksa-system",
+		Name:      vsphere.CredentialsObjectName,
+	}
+	if err := v.Client.Get(ctx, secretKey, secret); err != nil {
+		return nil, err
+	}
+	return secret, nil
+}
+
+func (v *VSphereReconciler) SetupEnvsAndDefaults(ctx context.Context, vsphereDatacenter *anywherev1.VSphereDatacenterConfig) error {
+	secret, err := v.VsphereCredentials(ctx)
+	if err != nil {
+		return fmt.Errorf("failed getting vsphere credentials secret: %v", err)
+	}
+
+	vsphereUsername := secret.Data["username"]
+	vspherePassword := secret.Data["password"]
+
+	if err := os.Setenv(vsphere.EksavSphereUsernameKey, string(vsphereUsername)); err != nil {
+		return fmt.Errorf("failed setting env %s: %v", vsphere.EksavSphereUsernameKey, err)
+	}
+
+	if err := os.Setenv(vsphere.EksavSpherePasswordKey, string(vspherePassword)); err != nil {
+		return fmt.Errorf("failed setting env %s: %v", vsphere.EksavSpherePasswordKey, err)
+	}
+
+	if err := vsphere.SetupEnvVars(vsphereDatacenter); err != nil {
+		return fmt.Errorf("failed setting env vars: %v", err)
+	}
+
+	if err := v.Defaulter.SetDefaultsForDatacenterConfig(ctx, vsphereDatacenter); err != nil {
+		return fmt.Errorf("failed setting default values for vsphere datacenter config: %v", err)
+	}
+
+	return nil
+}
+
+func (v *VSphereClusterReconciler) bundles(ctx context.Context, name, namespace string) (*releasev1alpha1.Bundles, error) {
+	clusterBundle := &releasev1alpha1.Bundles{}
+	err := v.capiResourceFetcher.FetchObjectByName(ctx, name, namespace, clusterBundle)
+	if err != nil {
+		return nil, err
+	}
+	return clusterBundle, nil
+}
+
+func (v *VSphereClusterReconciler) FetchAppliedSpec(ctx context.Context, cs *anywherev1.Cluster) (*c.Spec, error) {
+	return c.BuildSpecForCluster(ctx, cs, v.bundles, nil)
+}
+
+func (v *VSphereClusterReconciler) Reconcile(ctx context.Context, cluster *anywherev1.Cluster) (reconciler.Result, error) {
+	dataCenterConfig := &anywherev1.VSphereDatacenterConfig{}
+	dataCenterName := types.NamespacedName{Namespace: cluster.Namespace, Name: cluster.Spec.DatacenterRef.Name}
+	if err := v.Client.Get(ctx, dataCenterName, dataCenterConfig); err != nil {
```

Comment:
wonder if we should pass down the datacenterconfig fetched from https://github.com/aws/eks-anywhere/pull/921/files#diff-58f6d1beb0dffe9177008c7ea4959b2c4a4543ca561e8fe7b6cdf98114595272R86
instead of get it again using client

Reply:
The previous call has been removed